client/prometheus.go

```diff
@@ -0,0 +1,225 @@
+package client
+
+import (
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/hellofresh/stats-go/bucket"
+	"github.com/hellofresh/stats-go/incrementer"
+	"github.com/hellofresh/stats-go/state"
+	"github.com/hellofresh/stats-go/timer"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// Prometheus is Client implementation for prometheus
+type Prometheus struct {
+	sync.Mutex
+
+	unicode            bool
+	httpMetricCallback bucket.HTTPMetricNameAlterCallback
+	httpRequestSection string
+
+	namespace  string
+	increments map[string]*incrementer.Prometheus
+	states     map[string]*state.Prometheus
+	labels     map[string]string
+}
+
+// NewPrometheus builds and returns new Prometheus instance
+func NewPrometheus(namespace string) (*Prometheus, error) {
+	client := &Prometheus{
+		namespace:  namespace,
+		increments: make(map[string]*incrementer.Prometheus),
+		states:     make(map[string]*state.Prometheus),
+		labels:     make(map[string]string),
+	}
+	return client, nil
+}
+
+// WithLabels adds labels map to metrics
+func (c *Prometheus) WithLabels(labels map[string]string) Client {
+	c.Lock()
+	defer c.Unlock()
+
+	c.labels = labels
+	return c
+}
+
+// BuildTimer builds timer to track metric timings
+func (c *Prometheus) BuildTimer() timer.Timer {
+	return timer.NewPrometheus()
+}
+
+// Close closes underlying client connection if any
+func (c *Prometheus) Close() error {
+	return nil
+}
+
+// TrackRequest tracks HTTP Request stats
+func (c *Prometheus) TrackRequest(r *http.Request, t timer.Timer, success bool) Client {
+	b := bucket.NewHTTPRequest(c.httpRequestSection, r, success, c.httpMetricCallback, c.unicode)
+	metric := b.Metric()
+	metricTotal := b.MetricTotal()
+
+	// need to do smth with that
+	metric = strings.Replace(metric, "-.", "", -1)
+	metric = strings.Replace(metric, ".-", "", -1)
+	metric = strings.Replace(metric, "-", "", -1)
+	metric = strings.Replace(metric, ".", "_", -1)
+
+	metricTotal = strings.Replace(metricTotal, "-.", "", -1)
+	metricTotal = strings.Replace(metricTotal, ".-", "", -1)
+	metricTotal = strings.Replace(metricTotal, "-", "", -1)
+	metricTotal = strings.Replace(metricTotal, ".", "_", -1)
+
+	keyHash := ""
+
+	c.Lock()
+	for key := range c.labels {
+		keyHash += "_" + key
+	}
+
+	if _, ok := c.increments[metric+keyHash]; !ok {
+		c.increments[metric+keyHash] = incrementer.NewPrometheus(incrementer.NewPrometheusCounterFactory())
+	}
+
+	if _, ok := c.increments[metricTotal+keyHash]; !ok {
+		c.increments[metricTotal+keyHash] = incrementer.NewPrometheus(incrementer.NewPrometheusCounterFactory())
+	}
+
+	requestLabels := map[string]string{
+		"action":  r.Method,
+		"success": strconv.FormatBool(success),
+	}
+
+	c.increments[metric+keyHash].IncrementWithLabels(metric, requestLabels)
+	c.increments[metricTotal+keyHash].IncrementWithLabels(metricTotal, requestLabels)
+	c.labels = map[string]string{}
+	c.Unlock()
```

Comment:
defer this call right after locking, otherwise any potential problem (like panicing becuase of nil) will lock all requests